onnx/src/main/scala/vilcacora/onnx/Translator.scala

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright 2023 Arman Bilge
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package vilcacora.onnx
+
+import com.armanbilge.vilcacora.ir._
+import vilcacora.onnx.proto._
+import java.nio.ByteBuffer
+import java.nio.ByteOrder
+
+/** Translates an ONNX `ModelProto` into a custom, type-safe `ModelIR`.
+  *
+  * The primary goals of this translator are:
+  *   1. To convert the protobuf-based ONNX graph into a more easily consumable Scala ADT. 2. To
+  *      resolve all memory requirements at compile-time by creating `Allocation` objects for every
+  *      tensor (inputs, outputs, weights, and intermediate results). 3. To handle dynamic input
+  *      dimensions by substituting them with a fixed `maxBatchSize`.
+  */
+object Translator {
+
+  /** The main entry point for translation.
+    *
+    * @param model
+    *   The ONNX model loaded from a protobuf file.
+    * @param maxBatchSize
+    *   The static size to use for any dynamic dimensions in tensor shapes.
+    * @return
+    *   An `Either` containing the translated `ModelIR` on success, or an error message on failure.
+    */
+  def translate(model: ModelProto, maxBatchSize: Int = 1): Either[String, ModelIR] =
+    for {
+      graph <- model.graph.toRight("Model does not contain a graph")
+      allocations <- buildAllocations(graph, maxBatchSize)
+
+      // Translate all nodes into IR operations.
+
+      reversedOps <- graph.node.foldLeft[Either[String, List[Operation]]](Right(Nil)) {
+        (acc, node) =>
+          for {
+            opList <- acc
+            newOp <- translateNode(node)
+          } yield newOp :: opList
+      }
+      operations = reversedOps.reverse
```

Comment:
I think you may be able to write this more efficiently using `traverse`, so we don't have to reverse the list at the end.